RUN-3321 Move resource fetcher to use electron.net module

joneit opened the pull request:
**PLEASE HOLD OFF REVIEWING** as another commit is coming to handle `file:///`.

First commit is from #177. **Please restrict review comments for this PR to the other commits** (or hold off reviewing until I rebase it to develop after #177 is merged).

The final commit moves the internal API calls (`download-preload-scripts` and `get-preload-scripts`) from `System` to `Application`. Although `Window` would have been preferable (preload scripts are per window after all), the window doesn't exist yet so I put the APIs on `Application` instead.

I'm briefly caching each window's script set in `preloadCache` (a local in api/application.js)\*, keyed by uuid/name. This caching is very brief:
1. **before `loadUrl`:** window's set of script are downloaded asynchronously and cached
2. **after `loadUrl`** (at end of window spin-up): api-decorator synchronously executes each script in the set
3. **finally:** script set is released from memory cache after execution

[Tests](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/59c9481f7cb79f732d10c39b) passed, however... The new fetcher returns data directly to the caller (as opposed to a filepath in the cache folder to a file containing the data, as with the resource fetcher). This is nice as it simplifies the code, dispensing with the need to then open the files with `fs` module. This required structural changes to the other two consumers of resource fetcher, `Application.setTrayIcon` and `setTaskbarIcon` in api/window.js (the various "icon" window options), which I have yet to fully vet. That is, I need to set up some local tests for these to make sure they are working as expected. thoroughly.

\____________________
\* I welcome alternative suggestions for where to cache these.

joneit commented:
Tests are running now, however... The new fetcher returns data directly to the caller (as opposed to a filepath in the cache folder to a file containing the data, as with the resource fetcher). This required structure changes to the other two consumers of resource fetcher, `` and ``. I need to set up some local tests for these to make sure they are working as expected. That is to say, they may not pass automated tests, but even if they do I need to vet them more thoroughly.

rdepena commented on src/browser/clip_bounds.ts:
```diff
@@ -13,15 +13,14 @@ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * Clip width and height values to be within allowed maximum
  */
-function clipBounds(bounds: WindowBounds, browserWindow: BrowserWindow): WindowBounds {
+function clipBounds(bounds: Rectangle, browserWindow: BrowserWindow): Rectangle {
```
@joneit this is tangential to the code changes, revert it .

rdepena commented on src/browser/deferred.ts:
```diff
@@ -36,8 +35,8 @@ function handleMove(windowId: string, bounds: Bounds): void {
     if (isWin32 && browserWindow && (browserWindow.isMinimized() || browserWindow.isMaximized())) {
-        const oldBounds = browserWindow.getBounds();
-        const newBounds: WindowBounds = {
+        const oldBounds: Rectangle = browserWindow.getBounds();
```
tangential to the changes. revert.

rdepena commented on src/browser/log.ts:
```diff
@@ -33,25 +33,27 @@ export const logLevelMappings = new Map<any, any>([
 
-        if (typeof message === 'object') {
-            if (message instanceof Error) {
+        if (isObj && message instanceof Error) {
```
tangential to the changes. revert.

rdepena commented on src/modules.d.ts:
```diff
@@ -32,30 +33,71 @@ declare module 'electron' {
 
-    namespace BrowserWindow {
-        export function fromId(id: string): any;
+    export interface Rectangle {
```
tangential to the changes. revert.

rdepena commented on tslint.json:
```diff
@@ -86,7 +86,7 @@
         ],
         "no-console": [
-            true,
+            false,
```
revert this change.

rdepena commented on src/shapes.ts:
```diff
@@ -20,16 +20,6 @@ export interface Identity {
-    y: number;
-    width: number;
-    height: number;
-}
```
tangential, revert.

joneit commented:
Rebased to `8.56.25.3` and [retested](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/59c9cc087cb79f732d10c3af).

joneit commented:
Rebased -> #171 -> #169 -> canary (8.56.25.3) and [retested](https://testing-dashboard.openfin.co/#/app/sessions/api/completed/59c9df487cb79f732d10c3b1).

joneit commented:
Rebased to `8.56.25.8` + #177.
However, test runner gets `RVM exited` on all tests.... whatzupwitdat

joneit commented:
Rebased to `98.56.25.9`.
Retest pending...
Note that I still need to locally check `Application.setTrayIcon` and api/window.js::`setTaskbarIcon` as discussed in description above.

joneit commented:
**One more commit is coming (for `.ico` file support).**

Rebased to `8.56.25.11` and added 0f77dcd which:
* adds support for `file://` protocol (which `electron.net` does not support)
* resolves loading issues with:
   * `Application.setTrayIcon` (except for `.ico` files)
   * api/window.js::`setTaskbarIcon`
* [tested]() clean (ignore the 2 "preload" fails; they are resolved by #189)
